bin/propolis-server/src/lib/migrate/protocol.rs

```diff
@@ -0,0 +1,323 @@
+    NoEncodingVersionSeparator(String),
+
+    #[error("invalid encoding: {0}")]
+    InvalidEncoding(String),
+
+    #[error("failed to parse protocol version number {0}: {1}")]
+    InvalidVersionNumber(String, String),
+
+    #[error("offered protocol set contained duplicate protocol {0}")]
+    DuplicateProtocolInOffer(String),
+}
+
+/// The set of permissible encodings.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+enum Encoding {
+    /// Encode using Rust Object Notation.
+    Ron,
+}
+
+impl Display for Encoding {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Encoding::Ron => "ron",
+            }
+        )
+    }
+}
+
+impl FromStr for Encoding {
+    type Err = ProtocolParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "ron" => Ok(Encoding::Ron),
+            _ => Err(ProtocolParseError::InvalidEncoding(s.to_owned())),
+        }
+    }
+}
+
+/// A protocol selection.
+//
+// N.B. The ordering of fields in this struct matters! It ensures that the
+//      derived impls of PartialOrd and Ord compare versions before encodings.
+//      This ensures that the negotiation process always selects the latest
+//      version irrespective of whether it has the most preferable encoding.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+struct ProtocolParts {
+    version: u32,
+    encoding: Encoding,
+}
+
+impl ProtocolParts {
+    fn offer_string(&self) -> String {
+        format!(
+            "{PREFIX}{}{ENCODING_VERSION_SEPARATOR}{}",
+            self.encoding, self.version
+        )
+    }
+}
+
+impl From<Protocol> for ProtocolParts {
+    fn from(value: Protocol) -> Self {
+        match value {
+            Protocol::RonV0 => {
+                ProtocolParts { version: 0, encoding: Encoding::Ron }
+            }
+        }
+    }
+}
+
+impl FromStr for ProtocolParts {
+    type Err = ProtocolParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (encoding, version) = s
+            .strip_prefix(PREFIX)
+            .ok_or_else(|| ProtocolParseError::InvalidPrefix(s.to_owned()))?
+            .split_once(ENCODING_VERSION_SEPARATOR)
+            .ok_or_else(|| {
+                ProtocolParseError::NoEncodingVersionSeparator(s.to_owned())
+            })?;
+
+        let encoding = Encoding::from_str(encoding)?;
+        let version = version.parse().map_err(|e: ParseIntError| {
+            ProtocolParseError::InvalidVersionNumber(
+                version.to_owned(),
+                e.to_string(),
+            )
+        })?;
+
+        Ok(ProtocolParts { encoding, version })
+    }
+}
+
+/// The set of protocol variants supported by this migration library, expressed
+/// as a set of parts.
+static PROTOCOL_PARTS: Lazy<BTreeSet<ProtocolParts>> = Lazy::new(|| {
```

Comment:
Is `lazy_static` (as we've used elsewhere) inadequate for this?

Reply:
Nah. Fixed in 17dbf47.